src/pepr/operator/controllers/secrets/reload-utils.ts

```diff
@@ -0,0 +1,282 @@
+ */
+async function handleReplicaSetOwner(
+  namespace: string,
+  replicaSetName: string,
+  reason: string,
+  log: Logger,
+): Promise<void> {
+  try {
+    // Get the ReplicaSet
+    async function getReplicaSet() {
+      return K8s(kind.ReplicaSet).InNamespace(namespace).Get(replicaSetName);
+    }
+    const rs = await retryWithDelay(getReplicaSet, log);
+
+    // Look for a Deployment owner
+    const deploymentOwner = rs.metadata?.ownerReferences?.find(ref => ref.kind === "Deployment");
+
+    if (deploymentOwner?.name) {
+      // Found a Deployment owner, restart it
+      await restartController(namespace, kind.Deployment, deploymentOwner.name, reason, log);
+    } else {
+      // Standalone ReplicaSet - restart it directly using the same annotation pattern
+      await restartController(namespace, kind.ReplicaSet, replicaSetName, reason, log);
+    }
+  } catch (error) {
+    log.error(
+      { replicaSet: replicaSetName, namespace, error },
+      `Failed to handle ReplicaSet owner: ${reason}`,
+    );
+    throw error;
+  }
+}
+
+/**
+ * Restart a controller (Deployment, StatefulSet, DaemonSet, ReplicaSet) using the kubectl-style annotation
+ */
+export async function restartController(
+  namespace: string,
+  controllerKind: GenericClass,
+  name: string,
+  reason: string,
+  log: Logger,
+): Promise<void> {
+  // Get the controller kind name for logging
+  const controllerKindName = controllerKind?.name ?? String(controllerKind);
+
+  // Create a list of allowed controller kinds
+  const allowedKinds = [kind.Deployment, kind.StatefulSet, kind.DaemonSet, kind.ReplicaSet];
+
+  // Check if the provided kind is in our allowed list
+  // We use strict equality to check if it's the same object reference
+  const isAllowedKind = allowedKinds.some(k => k === controllerKind);
+
+  if (!isAllowedKind) {
+    throw new Error(`Unsupported controller kind: ${controllerKindName}`);
+  }
+
+  // Get the controller resource for the event
+  async function getController() {
+    return K8s(controllerKind).InNamespace(namespace).Get(name);
+  }
+  const controller = await retryWithDelay(getController, log);
+
+  try {
+    // Setup parent fields if any are missing to ensure our apply doesn't hit errors
+    if (!controller.spec) controller.spec = {};
+    if (!controller.spec.template) controller.spec.template = {};
+    if (!controller.spec.template.metadata) controller.spec.template.metadata = {};
+    if (!controller.spec.template.metadata.annotations)
+      controller.spec.template.metadata.annotations = {};
+    controller.spec.template.metadata.annotations["uds.dev/restartedAt"] = new Date().toISOString();
+    // Clear managedFields before apply
+    delete controller.metadata?.managedFields;
+
+    // Update the annotation in the controller object and apply
+    async function applyControllerAnnotation() {
+      return K8s(controllerKind, { name, namespace }).Apply(controller);
+    }
```

Comment:
Just to call this change out since the first iteration - a `Patch` on the controller would fail previously if the full parent path didn't exist (this is also true with `kubectl patch`). By using an apply and checking each parent path we ensure we won't hit unexpected errors here.